zio-http/src/test/scala/zhttp/http/HttpSpec.scala

```diff
@@ -334,6 +376,100 @@ object HttpSpec extends DefaultRunnableSpec with HExitAssertion {
+                case _: IllegalArgumentException => Http.succeed("bar")
+              }
+          assert(http.execute({}))(isSuccess(equalTo("bar")))
+        } +
+          test("keeps an error if doesn't catch anything") {
+            val exception = new Throwable("boom")
+            val http =
+              Http
+                .fail(exception)
+                .catchSome {
+                  case _: ArithmeticException => Http.succeed("bar")
+                }
+            assert(http.execute({}))(isFailure(equalTo(exception)))
+          } +
+          test("doesn't affect the success") {
+            val http =
+              (Http.succeed("bar") : Http[Any, Throwable, Any, String])
+                .catchSome {
+                  case _: Throwable => Http.succeed("baz")
+                }
+            assert(http.execute({}))(isSuccess(equalTo("bar")))
+          }
+      } +
+      suite("refineOrDie") {
+        test("refines matching exception") {
+          val http =
+            Http.fail(new IllegalArgumentException("boom"))
+              .refineOrDie {
+                case _: IllegalArgumentException => "fail"
+              }
+          assert(http.execute({}))(isFailure(equalTo("fail")))
+        } +
+          test("dies if doesn't catch anything") {
+            val t = new Throwable("boom")
+            val http =
+              Http
+                .fail(t)
+                .refineOrDie {
+                  case _: IllegalArgumentException => "fail"
+                }
+            assert(http.execute({}))(isDie(equalTo(t)))
+          } +
+          test("doesn't affect the success") {
+            val http =
+              (Http.succeed("bar") : Http[Any, Throwable, Any, String])
+                .refineOrDie {
+                  case _: Throwable => Http.succeed("baz")
+                }
+            assert(http.execute({}))(isSuccess(equalTo("bar")))
+          }
+      } +
+      suite("orDie") {
+        test("dies on failure") {
+          val t = new Throwable("boom")
+          val http =
+            Http.fail(t).orDie
+          assert(http.execute({}))(isDie(equalTo(t)))
+        } +
+          test("doesn't affect the success") {
+            val http =
+              (Http.succeed("bar") : Http[Any, Throwable, Any, String]).orDie
+            assert(http.execute({}))(isSuccess(equalTo("bar")))
+          }
+      } +
+      suite("catchSomeDefect") {
+        test("catches defect") {
+          val t = new IllegalArgumentException("boom")
+          val http = Http.die(t).catchSomeDefect { case _: IllegalArgumentException => Http.succeed("OK") }
+          assert(http.execute({}))(isSuccess(equalTo("OK")))
+        } +
+          test("propagates non-caught defect") {
+            val t = new IllegalArgumentException("boom")
+            val http = Http.die(t).catchSomeDefect { case _: SecurityException => Http.succeed("OK") }
+            assert(http.execute({}))(isDie(equalTo(t)))
+          }
```

Comment:
```suggestion
          } +
          test("catches thrown defects") {
            val t = new IllegalArgumentException("boom")
            val http = Http.collect({ _ => throw t}).catchSomeDefect { case _: IllegalArgumentException => Http.succeed("OK") }
            assert(http.execute({}))(isSuccess(equalTo("OK")))
          }
```

May be a test like this is required. Essentially a regression on #1021

Reply:
Done.